shared/shot.js

```diff
@@ -0,0 +1,726 @@
+  if (match) {
+    return match[0];
+  }
+  return null;
+}
+
+/** Check if the given object has all of the required attributes, and no extra
+    attributes exception those in optional */
+function checkObject(obj, required, optional) {
+  if (typeof obj != "object" || obj === null) {
+    throw new Error("Cannot check non-object: " + (typeof obj) + " that is " + JSON.stringify(obj));
+  }
+  required = required || [];
+  for (let attr of required) {
+    if (! (attr in obj)) {
+      return false;
+    }
+  }
+  optional = optional || [];
+  for (let attr in obj) {
+    if (required.indexOf(attr) == -1 && optional.indexOf(attr) == -1) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/** Create a JSON object from a normal object, given the required and optional
+    attributes (filtering out any other attributes).  Optional attributes are
+    only kept when they are truthy. */
+function jsonify(obj, required, optional) {
+  required = required || [];
+  let result = {};
+  for (let attr of required) {
+    result[attr] = obj[attr];
+  }
+  optional = optional || [];
+  for (let attr of optional) {
+    if (obj[attr]) {
+      result[attr] = obj[attr];
+    }
+  }
+  return result;
+}
+
+/** Resolve url relative to base */
+function resolveUrl(base, url) {
+  // FIXME: totally ad hoc and probably incorrect, but we can't
+  // use any libraries in this file
+  if (url.search(/^https?:/) != -1) {
+    // Absolute url
+    return url;
+  }
+  if (url.indexOf("//") === 0) {
+    // Protocol-relative URL
+    return (/^https?:/i).exec(base)[0] + url;
+  }
+  if (url.indexOf("/") === 0) {
+    // Domain-relative URL
+    return (/^https?:\/\/[a-z0-9\.\-]+/i).exec(base)[0] + url;
+  }
+  // Otherwise, a full relative URL
+  while (url.indexOf("./") === 0) {
+    url = url.substr(2);
+  }
+  let match = (/.*\//).exec(base)[0];
+  if (match.search(/^https?:\/$/i) === 0) {
+    // Domain without path
+    match = match + "/";
+  }
+  return match + url;
+}
+
+/** True if the two objects look alike.  Null, undefined, and absent properties
+    are all treated as equivalent.  Traverses objects and arrays */
+function deepEqual(a, b) {
+  if ((a === null || a === undefined) && (b === null || b === undefined)) {
+    return true;
+  }
+  if (typeof a != "object" || typeof b != "object") {
+    return a === b;
+  }
+  if (Array.isArray(a)) {
+    if (! Array.isArray(b)) {
+      return false;
+    }
+    if (a.length != b.length) {
+      return false;
+    }
+    for (let i=0; i<a.length; i++) {
+      if (! deepEqual(a[i], b[i])) {
+        return false;
+      }
+    }
+  }
+  if (Array.isArray(b)) {
+    return false;
+  }
+  let seen = {};
+  for (let attr in a) {
+    if (! deepEqual(a[attr], b[attr])) {
+      return false;
+    }
+    seen[attr] = true;
+  }
+  for (let attr in b) {
+    if (! seen[attr]) {
+      if (! deepEqual(a[attr], b[attr])) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function makeUuid() {
```

Comment:
I should probably rename this function, as it's really just a unique slug and not a UUID – in this particular instance it's just used to make unique IDs for multiple clips in a single shot.  To avoid collision we use random IDs in an object instead of putting them in an array.